nshlib/nsh_envcmds.c

```diff
@@ -533,29 +533,49 @@ int cmd_unset(FAR struct nsh_vtbl_s *vtbl, int argc, char **argv)
 {
 #if defined(CONFIG_NSH_VARS) || !defined(CONFIG_DISABLE_ENVIRON)
   int status;
+#endif
+#if defined(CONFIG_NSH_VARS) && !defined(CONFIG_DISABLE_ENVIRON)
+  FAR char *oldvalue;
 #endif
   int ret = OK;
 
 #if defined(CONFIG_NSH_VARS)
-  /* Unset NSH variable */
+#ifndef CONFIG_DISABLE_ENVIRON
+  /* Check if the NSH variable has already been promoted to an group-
+   * wide environment variable. In this case, remove the exported
+   * variable
+   */
 
-  status = nsh_unsetvar(vtbl, argv[1]);
-  if (status < 0 && status != -ENOENT)
+  oldvalue = getenv(argv[1]);
+  if (oldvalue == NULL)
+#endif
     {
-      nsh_error(vtbl, g_fmtcmdfailed, argv[0], "nsh_unsetvar",
-                NSH_ERRNO_OF(-status));
-      ret = ERROR;
+      /* Unset NSH variable */
+
+      status = nsh_unsetvar(vtbl, argv[1]);
+      if (status < 0 && status != -ENOENT)
+        {
+          nsh_error(vtbl, g_fmtcmdfailed, argv[0], "nsh_unsetvar",
+                    NSH_ERRNO_OF(-status));
+          ret = ERROR;
+        }
     }
 #endif
 
 #if !defined(CONFIG_DISABLE_ENVIRON)
-  /* Unset environment variable */
-
-  status = unsetenv(argv[1]);
```

Comment:
it's more simple to skip the error message if  errno set by unsetenv equals ENOENT.

Reply:
Agreed. Maybe wr can assume that the variable doesn't exist in both nsh_vars and environment variables (i.e. Local variable can't shadow an environment variable). Since in that case the implementation would be wrong; unsetting a local variable (in script) would also delete the global one. This was the thinking behind, why I wrote it like that. But I can change it as you suggested for sure!

Reply:
Event so, We can simplify the code like this:
```
status = nsh_unsetvar(vtbl, argv[1]);
#if !defined(CONFIG_DISABLE_ENVIRON)
if (status == -ENOENT)
  {
     status = unsetenv(argv[1]);
      if (status < 0)
        {
          nsh_error(vtbl, g_fmtcmdfailed, argv[0], "unsetenv",
                    NSH_ERRNO);
          ret = ERROR;
        }
  }
else if (status < 0)
  {
      nsh_error(vtbl, g_fmtcmdfailed, argv[0], "nsh_unsetvar",
                NSH_ERRNO_OF(-status));
      ret = ERROR;
  }
#endif
```

Reply:
Looks good to me, go ahead and make the PR, as you have the gode ready.